### PR TITLE
#166272336 Feature Sign out user

### PR DESCRIPTION
**Pull request by despeauxz**
#### What does this PR do?
This PR adds functionality for a user to sign out 
#### Description of Task to be completed?
- Created a DroppedToke table for dropped tokens
- Create an authorize middleware that checks the token if it has been dropped or is not provided
#### How should this be manually tested?
Navigate to the endpoint `http://localhost:3000/api/v1/auth/logout`
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166272336](https://www.pivotaltracker.com/story/show/166272336)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

### CODE REVIEW

File: server/middlewares/Authorization.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+class Authorization {
+  /**
+    * Get token from req
+    * @param  {object} req - Request object
+    * @returns {string} token
+    * @static
+    */
+  static getToken(req) {
+    const bearerToken = req.headers.authorization;
+    const token = bearerToken && bearerToken.replace('Bearer ', '');
```
Comment (fidelisojeah) [inline server/middlewares/Authorization.js]: token would be `false` if `bearerToken` does not exist

You should properly change this to something like:

```javascript
if (bearerToken && bearerToken.startsWith('Bearer ')){
  const token = bearerToken.replace('Bearer ', '');
  return token
} else if (bearerToken){
  return bearerToken; // This part is not entirely needed (only gives the option for the token to be set in the header directly
}else {
  throw new Error('Invalid token supplied');
}
```

then move the `const token = Authorization.getToken(req);` into the `try/catch` block

**Review by fidelisojeah — CHANGES_REQUESTED** [review]

File: server/middlewares/Authorization.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    * @async
+    * @param  {object} req - Request object
+    * @param {object} res - Response object
+    * @param {object} next The next middleware
+    * @return {json} Returns json object
+    * @static
+    */
+  static async authorize(req, res, next) {
+    const token = Authorization.getToken(req);
+    if (!token) return res.status(400).json({ status: 400, error: 'No token specified' });
```
Comment (fidelisojeah) [inline server/middlewares/Authorization.js]:
```javascript
errors: {
  global: 'Invalid token supplied: format Bearer <token>'
}
```

**Review by fidelisojeah — CHANGES_REQUESTED** [review]

File: server/controllers/users.js [hunk 3]
```diff
@@ -150,6 +150,27 @@ class UserController {
...
+   */
+  static async logout(req, res) {
+    const token = await Token.getToken(req);
+    try {
+      await DroppedToken.create({ token });
+      return res.status(201).json({
+        status: 201, message: 'You are now logged out'
+      });
+    } catch (error) {
+      return res.status(401).json({
```
Comment (fidelisojeah) [inline server/controllers/users.js]: similar error format issue here

Change the error format here

File: server/helpers/Token.js [hunk 4]
```diff
@@ -23,27 +26,56 @@ class Token {
...
+        return Response.error(res, 401, 'Invalid Token Provided');
+      }
+
+      const droppedToken = await DroppedToken.findOne({
+        where: {
+          token
+        }
+      });
+      if (droppedToken) {
+        return res.status(401).json({
```
Comment (fidelisojeah) [inline server/helpers/Token.js]: wrong error format

File: server/models/droppedtoken.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+module.exports = (sequelize, DataTypes) => {
+  const DroppedToken = sequelize.define('DroppedToken', {
+    token: DataTypes.TEXT,
+    createdAt: DataTypes.DATE,
+    updatedAt: DataTypes.DATE
+  }, {});
+
+  // DroppedToken.associate = (models) => {
```
Comment (fidelisojeah) [inline server/models/droppedtoken.js]: remove commented code

**Review by fidelisojeah — CHANGES_REQUESTED** [review]
Please pull from develop and correct current issues

**Comment by GerrardE** [comment]
Good job. Do you mind updating the API docs?

**Review by EMacco — CHANGES_REQUESTED** [review]
Great work Malik, although I think should could also consider handling the error if a token is not provided at all. You could return a global error `You are not logged in`. Currently it returns `Cannot read property 'startsWith' of undefined`
Screenshot 2019-06-11 at 10 47 05 PM

Also, considering the fact that you now have Bearer appended to the token, you could update the login and register routes so that it returns the token with 'Bearer ' appended to it. 
Screenshot 2019-06-11 at 10 49 40 PM

**Comment by fidelisojeah** [comment]
> Great work Malik, although I think should could also consider handling the error if a token is not provided at all. You could return a global error `You are not logged in`. Currently it returns `Cannot read property 'startsWith' of undefined`
> Screenshot 2019-06-11 at 10 47 05 PM
> 
> Also, considering the fact that you now have Bearer appended to the token, you could update the login and register routes so that it returns the token with 'Bearer ' appended to it.
> Screenshot 2019-06-11 at 10 49 40 PM


please do not append `Bearer ` to the token on generation

However, all other feedback is okay

**Review by despeauxz — COMMENTED** [review]
@fidelisojeah The story does not require I document the endpoint with swagger and the only field required is the req.header.authorization. As for the other feedback by @EMacco, that doesn't apply cos I just tested again to be sure and no such error. But will definitely update the error response feedback

**Review by EMacco — APPROVED** [review]

**Review by GerrardE — APPROVED** [review]

**Review by fidelisojeah — APPROVED** [review]

**Pull request merged**